### PR TITLE
Add failing test case for issue #549 (empty Importance object print)

### LINKED ISSUE #549: Blank importance objects can't be printed
**Describe the bug**

Importance is missing a default `_tree` and when `str` is called this leads to an `AttributeError`.

**To Reproduce**

A short code snippet of what you have ran. Please change or remove any specific values or anything that can't be public. For example:

``` python
problem = montepy.read_input("tests/inputs/test_importance.imcnp")
print(problem.cells._importance)

```

**Error Message (if any)**

If an error message was printed please include the entire stacktrace. If it includes any specific values please change or remove them. For example:

``` python
    print(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Importance: in_cell_block: False, set_in_cell_block False,
{}

    def __str__(self):
        if not self.in_cell_block and self._problem is None:
>           return " ".join(self.input_lines)
E           AttributeError: 'Importance' object has no attribute 'input_lines'

montepy/data_inputs/importance.py:165: AttributeError
```

**MCNP input file snippet**

see `tests/inputs/test_importance.imcnp`

**Version**

 - Version 0.4.1

### PR DESCRIPTION
**Pull request by kordusas**
# Pull Request Checklist for MontePy

### Description

Previously Importance class had a place holder for printing and failed on empty object instance. Now Importance printing is formatted like this: `IMPORTANCE: neutron=1.0, photon=1.0`

Empty object instance printing looks like this: `IMPORTANCE: Object is empty`

Reformated the _test_importance.py_ now using @pytest.mark.parametrize and removed dependence on TestCase object. 
* `test_importance_parsing_from_cell`  checks if valid values are initiated or valid error is raised when creating Cell instance and getting the importance instance from the  
* `test_importance_init_data_valid` tests __init__ of importance from DATA block and testing access to it
* `test_importance_init_data_invalid` - testing invalid importance object instance __init__
* added fixtures for `cell_with_importance `, `empty_cell`, `test_importance_values` to reuse same  cells and same particle and importance assingments 
* `test_str_repr_parsed_cells`  checks that methods return valid type
* `test_str_repr_empty_importance` checks that empty importance object returns whats expected 
* `test_importance_manual_assignment_and_str_repr` - tests manual assignment of importance and the correct strrepresentation of the importance instance

Fixes #549

---

### General Checklist

- [x] I have performed a self-review of my own code.
- [x] The code follows the standards outlined in the [development documentation](https://idaholab.github.io/MontePy/developing.html).
- [x] I have formatted my code with `black` version 25.
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable).

---

<details open> 

<summary><h3>Documentation Checklist</h3></summary>

- [ ] ~I have documented all added classes and methods.~
- [ ] ~For infrastructure updates, I have updated the developer's guide.~
- [ ] ~For significant new features, I have added a section to the getting started guide.~

</details>

---

<details open>
<summary><h3>First-Time Contributor Checklist</h3></summary>

- [x] If this is your first contribution, add yourself to `pyproject.toml` if you wish to do so.

</details>

---

### Additional Notes for Reviewers

Ensure that:

- [x] The submitted code is consistent with the merge checklist outlined [here](https://www.montepy.org/developing.html#merge-checklist).
- [x] The PR covers all relevant aspects according to the development guidelines.
- [x] 100% coverage of the patch is achieved, or justification for a variance is given.

----
📚 Documentation preview 📚: https://montepy--681.org.readthedocs.build/en/681/

### CODE REVIEW

**Review by MicahGale — CHANGES_REQUESTED** [review]
Thanks for the contribution @kordusas!

This is a good first pass but some changes are needed. I left comments on specific changes.

1. Update the PR description to give a summary of what the problem that was solved is, and how it was solved. Also link to the specific issues. See #680 for an example. 
2. Do you want to add yourself as an author to `AUTHORS` and/or `pyproject.toml`?
3. Make sure to update the changelog with this bug fix.
4. Would you like to added as a contributor to this repo on GH?

File: montepy/data_inputs/importance.py [hunk 1]
```diff
@@ -174,10 +174,31 @@ def __delitem__(self, particle):
...
+        ret = []
+        for particle, tree in self._particle_importances.items():
+            # Instead of tree["classifier"].particles.value (which doesn't exist),
+            # use str(tree["classifier"].particles) or an appropriate attribute.
+            ret.append(
+                f"{tree['classifier'].prefix.value}"
+                f"{str(tree['classifier'].particles)}"
+                f"{tree['seperator'].value}"
+                f"{tree['data'].nodes[0].value}"
+            )
```
[the remark this answers is not in the document]
Comment (kordusas) [reply montepy/data_inputs/importance.py:188]: yes, thank you, need to get used to multiple inheritances here.

**Comment by MicahGale** [comment]
@kordusas do you need any help with this PR?

File: montepy/data_inputs/importance.py [hunk 2]
```diff
@@ -174,10 +174,31 @@ def __delitem__(self, particle):
...
-        return "".join(self.format_for_mcnp_input(DEFAULT_VERSION))
+        if self._problem:
+            # Use the full formatting (which relies on _problem.cells, etc.)
+            return "".join(self.format_for_mcnp_input(DEFAULT_VERSION))
+        else:
+            # Fall back to a default simple representation 
+            default = self._format_default()
+            if default:
+                return " ".join(default)
+            return "Importance object is empty"
```
[the remark this answers is not in the document]
Comment (kordusas) [reply montepy/data_inputs/importance.py:198]: maybe something like this then?  printing out particle name explicitly might be more user intuitive. 
```

print(imp)
IMPORTANCE: neutron=1.0 photon=1.0

```
* is there an intentional purpose of :  `' '.join(self.format_for_mcnp_input(DEFAULT_VERSION))` in here? 

*And what should be printed if `problem.print_in_data_block["IMP"] = True` . I think even in this case it should print the values of the selected cell

File: montepy/data_inputs/importance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kordusas) [reply montepy/data_inputs/importance.py]: It does lead to duplicate values, but i think this could be sufficient here: 
` ret.append( f"{particle}={tree['data'].nodes[0].value}")`

File: montepy/data_inputs/importance.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MicahGale) [reply montepy/data_inputs/importance.py:198]: > ```
> print(imp)
> IMPORTANCE: neutron=1.0 photon=1.0
> ```

Nitpick: I think commas would be nice.

```
print(imp)
IMPORTANCE: neutron=1.0, photon=1.0
```

> is there an intentional purpose of : ' '.join(self.format_for_mcnp_input(DEFAULT_VERSION)) in here?

This comes back to running `format_for_mcnp_input` was a fast and dirty implement to have _something_ for `__str__`. The goal really should be to get away from it all together. The reason it's called that way though is because it returns `list[str]` so `''.join` is a fast way to make it into a single string. 

> What should be printed if problem.print_in_data_block["IMP"] = True . I think even in this case it should print the values of the selected cell. Currently the print is blank if its True and problem exists.

I agree what `str` shows should be independent of `problm.print_in_data_block` because otherwise this will be _very confusing to users and look like a bug_ (ask me how I know).

File: tests/test_importance.py [hunk 3]
```diff
@@ -31,6 +32,10 @@ def test_importance_init_cell(self):
         with self.assertRaises(ValueError):
             cell = Cell(card)
 
+    def test_importance_init_empty(self):
+        importance = Importance()
+        assert str(importance) == "Importance object is empty"
+
```
[the remark this answers is not in the document]
Comment (kordusas) [reply tests/test_importance.py]: Hi, sorry for delay, I think I implemented all suggestions and tried to follow the recomendations. 
I used some parametrisation in the test_importance, i think this adds flexibility of easily adding new input parsing cases to test. But let me know if you think this should be rewritten. 

It may be good to refactor the importance testing further, if you have suggestions let me know. 

Also was wondering is the `IMP:N,P 1 0` input acceptable when `mcnp_input.Input([in_str], block_type.BlockType.CELL)` isn't this explicitly DATA block input format?  

Let me know if I didn't follow some of the guidelines. Hopefully this was helpful. 

cheers

File: doc/source/changelog.rst [hunk 4]
```diff
@@ -5,6 +5,12 @@ MontePy Changelog
 0.5 releases
 ============
 
+0.5.6
```
Comment (MicahGale) [inline doc/source/changelog.rst:8]:
``` suggestion
#Next Version#
```

We have a script that changes the version number at the last minute. (Left over from when we are trying to go all in on CD).

File: montepy/data_inputs/importance.py [hunk 5]
```diff
@@ -174,31 +174,24 @@ def __delitem__(self, particle):
...
         else:
-            # Fall back to a default simple representation 
-            default = self._format_default()
-            if default:
-                return " ".join(default)
-            return "Importance object is empty"
-
+            return "IMPORTANCE: Object is empty"
 
     def __repr__(self):
```
Comment (MicahGale) [inline montepy/data_inputs/importance.py:196]: Has this bug been tested for `print(repr(...))` too?

File: montepy/data_inputs/importance.py [hunk 6]
```diff
@@ -174,31 +174,24 @@ def __delitem__(self, particle):
             raise TypeError("Key must be a particle")
         del self._particle_importances[particle]
 
-    def _format_default(self):
-        # Create a simple, self-contained string representation of the importance settings.
+    def _format_default(self) -> list[str]:
```
Comment (MicahGale) [inline montepy/data_inputs/importance.py:177]: Is this function called anywhere else? If not it's small enough that it could just moved inside of `__str__`

File: tests/test_importance.py [hunk 7]
```diff
@@ -1,189 +1,328 @@
 # Copyright 2024, Battelle Energy Alliance, LLC All Rights Reserved.
-from unittest import TestCase
 import montepy
+import os, pytest
```
Comment (MicahGale) [inline tests/test_importance.py:3]: See the [new style guide for imports (in alpha release)](https://www.montepy.org/en/v1.0.0a4/dev_standards.html#style-guide). We don't say anything about this but we have always done one package per line. @tjlaboss do you have thoughts on multi-package imports?

File: tests/test_importance.py [hunk 8]
```diff
@@ -1,189 +1,328 @@
...
 from montepy.cell import Cell
 from montepy.particle import Particle
 from montepy.data_inputs.importance import Importance
 from montepy.errors import *
 from montepy.input_parser import mcnp_input, block_type
-import os
-from montepy.data_inputs.importance import Importance
```
Comment (MicahGale) [inline tests/test_importance.py]: Lol. Whoops. This was definitely my fault.

File: tests/test_importance.py [hunk 9]
```diff
@@ -1,189 +1,328 @@
...
-            cell = Cell(card)
-        # test negative imp
-        in_str = "1 0 -1 IMP:N,P=-2"
-        card = mcnp_input.Input([in_str], block_type.BlockType.CELL)
-        with self.assertRaises(ValueError):
-            cell = Cell(card)
+def create_cell_from_input(in_str, block=block_type.BlockType.CELL):
+    """Helper to create a Cell object from a given input string."""
+    card = mcnp_input.Input([in_str], block)
+    return Cell(card)
```
Comment (MicahGale) [inline tests/test_importance.py:14]: FYI this can be removed in `1.0.0`. `Importance(in_str)` is now supported in that version.

File: tests/test_importance.py [hunk 10]
```diff
@@ -1,189 +1,328 @@
...
+            None,
+        ),  # default neutron importance when nothing is set
+        # Error cases
+        ("1 0 -1 IMP:N,P=h", None, ValueError),  # non-numeric value
+        ("1 0 -1 IMP:N,P=-2", None, ValueError),  # negative value
+        ("1 0 -1 IMP:N,xx=2", None, ParsingError),  # invalid particle type
+    ],
+)
+def test_importance_parsing_from_cell(in_str, expected, error):
+    """Test importance parsing from cell input string.
```
Comment (MicahGale) [inline tests/test_importance.py:50]: Thanks for the doc strings. `tests` are generally exempt from this expectation.

File: tests/test_importance.py [hunk 11]
```diff
@@ -1,189 +1,328 @@
...
+    ],
+)
+def test_importance_init_data_invalid(in_str, kwargs, expected_exception):
+    card = mcnp_input.Input([in_str], block_type.BlockType.DATA)
+    with pytest.raises(expected_exception):
+        Importance(card, **kwargs)
+
+
+class TestImportance:
+
```
Comment (MicahGale) [inline tests/test_importance.py:123]: Why are the previous tests not inside of this class? To me they are all related. See this [test](https://github.com/idaholab/MontePy/blob/alpha-test/tests/test_material.py#L96) for how to use parametrize inside of a class. (I use `_` instead of `self` for brevity).

File: tests/test_importance.py [hunk 12]
```diff
@@ -1,189 +1,328 @@
...
-        problem.mode.add(montepy.particle.Particle.PHOTON)
+        problem.mode.add(Particle.NEUTRON)
+        problem.mode.add(Particle.PHOTON)
         imp = cell.importance
         cell.link_to_problem(problem)
         imp.all = 2.0
-        self.assertAlmostEqual(imp.neutron, 2.0)
-        self.assertAlmostEqual(imp.photon, 2.0)
+        assert imp.neutron == 2.0
+        assert imp.photon == 2.0
```
Comment (MicahGale) [inline tests/test_importance.py:228]:
``` suggestion
        assert imp.neutron == pytest.approx(2.0)
        assert imp.photon == pytest.approx(2.0)
```

Yes this is low risk for rounding error but still you should never evaluate equality directly with floats.

File: tests/test_importance.py [hunk 13]
```diff
@@ -1,189 +1,328 @@
...
+        """
+        cell = cell_with_importance
+        # Test setting first value from test values
+        particle, value = next(iter(test_importance_values.items()))
+        cell.importance[particle] = value
+        assert cell.importance[particle] == value
+
         cell.importance.neutron = 2.5
-        self.assertEqual(cell.importance.neutron, 2.5)
+        assert cell.importance.neutron == 2.5
```
Comment (MicahGale) [inline tests/test_importance.py:250]: `pytest.approx` again.

File: tests/test_importance.py [hunk 14]
```diff
@@ -1,189 +1,328 @@
...
+        """
+        Test deletion of importance values.
+
+        Args:
+            parsed_cells: Fixture providing cells with importance assignments
+        """
+        cell = cell_with_importance
         del cell.importance.neutron
-        self.assertAlmostEqual(cell.importance.neutron, 0.0)
+        assert cell.importance.neutron == 0.0
```
Comment (MicahGale) [inline tests/test_importance.py:281]: `pytest.approx` again.

File: tests/test_importance.py [hunk 15]
```diff
@@ -1,189 +1,328 @@
...
         del cell.importance[Particle.PHOTON]
-        self.assertAlmostEqual(cell.importance.photon, 0.0)
-        with self.assertRaises(TypeError):
+        assert cell.importance.photon == 0.0
```
Comment (MicahGale) [inline tests/test_importance.py:283]: `pytest.approx`

File: tests/test_importance.py [hunk 16]
```diff
@@ -1,189 +1,328 @@
...
-        self.assertEqual(
-            [
-                val.value
-                for val in imp1._particle_importances[Particle.ELECTRON]["data"]
-            ],
-            [0.0, 0.0],
-        )
+        assert [
+            val.value for val in imp1._particle_importances[Particle.NEUTRON]["data"]
+        ] == [1.0, 0.0]
```
Comment (MicahGale) [inline tests/test_importance.py:301]: TIL `pytest.approx` can be used with lists.

**Review by MicahGale — CHANGES_REQUESTED** [review]
I like what you have done with this `__str__` style. 

Also thank you for handling the conversion of `unittest->pytest` for this test suite, it wasn't essential for this PR, but is a huge help.

File: montepy/data_inputs/importance.py (hunk 5 above)
Comment (MicahGale) [reply montepy/data_inputs/importance.py:196]: This is tested.

File: tests/test_importance.py (hunk 13 above)
Comment (kordusas) [reply tests/test_importance.py:250]: Thanks for pointing that out, I should not have left this like that

**Comment by MicahGale** [comment]
FYI: this will probably be deployed in 1.0.0, and not 0.5.6... hopefully. I'm hoping that we are really close to completing 1.0.0 release.

File: tests/test_importance.py (hunk 7 above)
Comment (kordusas) [reply tests/test_importance.py:3]: we can follow the established convention, this doesnt really make a lot of difference i think.

**Comment by kordusas** [comment]
> FYI: this will probably be deployed in 1.0.0, and not 0.5.6... hopefully. I'm hoping that we are really close to completing 1.0.0 release.

Should i wait for the 1.0.0 to come out before submitting another pull request?

File: tests/test_importance.py (hunk 11 above)
Comment (kordusas) [reply tests/test_importance.py:123]: As they are independant from the `TestImportance` class and not using any class variables or fixtures it seemed like it is redundant to add it into the class. Just gonna make calss big, thus i kept them separate. Whats the reasoning behind moving them into the class? because they are handling tests for importance?

**Comment by MicahGale** [comment]
> > FYI: this will probably be deployed in 1.0.0, and not 0.5.6... hopefully. I'm hoping that we are really close to completing 1.0.0 release.
> 
> Should i wait for the 1.0.0 to come out before submitting another pull request?

Handling the merge conflicts wouldn't be the biggest deal. What you can do though is make new branches off of `alpha-test-dev` instead of `develop`.

File: tests/test_importance.py (hunk 11 above)
Comment (MicahGale) [reply tests/test_importance.py:123]: Now, we aren't really using classes for infrastructure, but rather for logically grouping tests. For this file it's not a big deal. `tests/test_universe.py` though tests `Universe_input`, `Fill`, and `Lattice`, so having classes makes it a lot easier to find the relevant tests are they are grouped by what class they are testing. With a file that's testing one object I would say commit to one approach. Either everything is inside of that class, or there is no class. (fixtures can be at the module/file level).

File: doc/source/changelog.rst (hunk 4 above)
Comment (MicahGale) [reply doc/source/changelog.rst:74]: Please move this to the `1.0 releases` section since you rebased onto `alpha-test-dev`

**Review by MicahGale — APPROVED** [review]
This is looking good. Just one small tweak. Thanks for the contribution. If you have permissions you can merge when you are ready (we prefer merging over other options).
